File: 프로그래머스/2/42587. 프로세스/프로세스.java
import java.util.Arrays;
import java.util.Deque;
import java.util.LinkedList;

class Solution {
    public int solution(int[] priorities, int location) {
        Deque<Integer[]> queue = new LinkedList<>();
        Integer[] copy = new Integer[priorities.length];
        
        for (int i = 0; i < priorities.length; i++) {
            int priority = priorities[i];
            queue.add(new Integer[] {i, priority});
            copy[i] = priority;
        }
        
        
        Arrays.sort(copy, (priority1, priority2) -> priority2 - priority1);
        
        int pointer = 0;
        int time = 0;
        
        while (!queue.isEmpty()) {
            Integer[] process = queue.removeFirst();
            if (process[1] != copy[pointer]) {
                queue.add(process);
                continue;
            }
            time++;
            pointer++;
            if (process[0] == location) {
                return time;
            }
        }
        
        return -1;
    }
}

File: 프로그래머스/2/42587. 프로세스/프로세스.kt
import java.util.PriorityQueue

data class Process(val index: Int,val priority: Int)

class Solution {
    fun solution(priorities: IntArray, location: Int): Int {
        val pq = PriorityQueue<Int>(priorities.size, compareBy { -it })
        pq.addAll(priorities.asList())
        
        val processes = priorities.withIndex().map {Process(it.index, it.value)}
        val queue = ArrayDeque<Process>(processes)
        var answer = 0
        
        while (queue.isNotEmpty()) {
            val process = queue.removeFirst()
            val (index, priority) = process
            
            if (priority != pq.element()) {
                queue.add(process)
                continue;
            }
            
            pq.remove()
            answer++
            
            if (index == location) {
                return answer
            }
        }
        
        return answer
    }
}